File: lang/ruby/src/fun/vac/data_structures/LinkedList.rb
module LinkedList

    class Node

        attr_accessor :data
        attr_accessor :next

        public
        def initialize(element)
            @data = element
            @next = nil
        end
    end

    class LinkedList

        public
        def initialize()
            @head = nil
            @tail = nil
            @length = 0
        end

        public
        def size()
            @length
        end

        def isEmpty()
            @length.zero?
        end

        def get(index)
            if index.negative? || index >= @length
                raise "[PANIC - IndexOutOfBounds]"
            end

            cursor = nil

            if index == @length - 1
                cursor = @tail
            else
                cursor = @head
                (0 ... index).each do
                    cursor = cursor.next
                end
            end

            cursor.data
        end

        def set(index, element)
            if index.negative? || index >= @length
                raise "[PANIC - IndexOutOfBounds]"
            end

            cursor = nil

            if index == @length - 1
                cursor = @tail
            else
                cursor = @head
                (0 ... index).each do
                    cursor = cursor.next
                end
            end

            cursor.data = element
        end

        def insert(index, element)
            if index.negative? || index > @length
                raise "[PANIC - IndexOutOfBounds]"
            end

            node = Node.new(element)

            case index
            when 0
                if @length != 0
                    node.next = @head
                else
                    @tail = node
                end
                @head = node
            when @length
                @tail.next = node
                @tail = node
            else
                cursor = @head
                (0 ... index - 1).each do
                    cursor = cursor.next
                end
                node.next = cursor.next
                cursor.next = node
            end

            @length += 1
        end

        def remove(index)
            if index.negative? || index >= @length
                raise "[PANIC - IndexOutOfBounds]"
            end

            target = nil

            if index.zero?
                target = @head
                if @length == 1
                    @head = nil
                    @tail = nil
                else
                    @head = target.next
                end
            else
                cursor = @head
                (0 ... index - 1).each do
                    cursor = cursor.next
                end
                target = cursor.next
                cursor.next = target.next
                if index == @length - 1
                    @tail = cursor
                end
            end

            target.data = nil

            @length -= 1
        end

        def front()
            get(0)
        end

        def back()
            get(@length - 1)
        end

        def prepend(element)
            insert(0, element)
        end

        def append(element)
            insert(@length, element)
        end

        def poll()
            remove(0)
        end

        def eject()
            remove(@length - 1)
        end
    end
end

File: lang/php/src/fun/vac/data_structures/LinkedList.php
<?php
    namespace fun\vac\data_structures\LinkedList;

    require_once "fun/vac/data_structures/interfaces/IList.php";

    use fun\vac\data_structures\interfaces\IList;

    class Node {
        public $data;
        public $next = NULL;

        public function __construct(int $element) {
            $this->data = $element;
        }
    }

    class LinkedList implements IList\IList {
        private $head = NULL;
        private $tail = NULL;
        private $length = 0;

        public function __construct() {}

        public function size(): int {
            return $this->length;
        }

        public function isEmpty(): bool {
            return $this->length == 0;
        }

        public function get(int $index): int {
            if ($index < 0 || $index >= $this->length) {
                throw new RuntimeException("[PANIC - IndexOutOfBounds]");
            }

            $cursor = NULL;

            if ($index == $this->length - 1) {
                $cursor = $this->tail;
            } else {
                $cursor = $this->head;
                for ($i = 0; $i < $index; $i++) {
                    $cursor = $cursor->next;
                }
            }

            return $cursor->data;
        }

        public function set(int $index, int $element): void {
            if ($index < 0 || $index >= $this->length) {
                throw new RuntimeException("[PANIC - IndexOutOfBounds]");
            }

            $cursor = NULL;

            if ($index == $this->length - 1) {
                $cursor = $this->tail;
            } else {
                $cursor = $this->head;
                for ($i = 0; $i < $index; $i++) {
                    $cursor = $cursor->next;
                }
            }

            $cursor->data = $element;
        }

        public function insert(int $index, int $element): void {
            if ($index < 0 || $index > $this->length) {
                throw new RuntimeException("[PANIC - IndexOutOfBounds]");
            }

            $node = new Node($element);

            switch ($index) {
            case 0:
                if ($this->length != 0) {
                    $node->next = $this->head;
                } else {
                    $this->tail = $node;
                }
                $this->head = $node;
                break;
            case $this->length:
                $this->tail->next = $node;
                $this->tail = $node;
                break;
            default:
                $cursor = $this->head;
                for ($i = 0, $bound = $index - 1; $i < $bound; $i++) {
                    $cursor = $cursor->next;
                }
                $node->next = $cursor->next;
                $cursor->next = $node;
            }

            $this->length++;
        }

        public function remove(int $index): void {
            if ($index < 0 || $index >= $this->length) {
                throw new RuntimeException("[PANIC - IndexOutOfBounds]");
            }

            $target = NULL;

            if ($index == 0) {
                $target = $this->head;
                if ($this->length == 1) {
                    $this->head = NULL;
                    $this->tail = NULL;
                } else {
                    $this->head = $target->next;
                }
            } else {
                $cursor = $this->head;
                for ($i = 0, $bound = $index - 1; $i < $bound; $i++) {
                    $cursor = $cursor->next;
                }
                $target = $cursor->next;
                $cursor->next = $target->next;
                if ($index == $this->length - 1) {
                    $this->tail = $cursor;
                }
            }

            $target->data = NULL;

            $this->length--;
        }

        public function front(): int {
            return $this->get(0);
        }

        public function back(): int {
            return $this->get($this->length - 1);
        }

        public function prepend(int $element): void {
            $this->insert(0, $element);
        }

        public function append(int $element): void {
            $this->insert($this->length, $element);
        }

        public function poll(): void {
            $this->remove(0);
        }

        public function eject(): void {
            $this->remove($this->length - 1);
        }
    }
?>
